### PR TITLE
Add push module

### PR DESCRIPTION
**Pull request by MatkovIvan**
Added push module to SDK.
- Send push token in push installation log
- Added module settings in sasquatch app UI

### CODE REVIEW

**Comment by jaelim-ms** [comment]
Please fix the broken build first. Sasquatch application has two flavors in build. One should be fine but the other shouldn't because it has dependencies with libraries that are in jcenter. You are introducing a new feature but it is not in jcenter yet. I recommend you to remove all Sasquatch implementation at the moment.

**Comment by guperrot** [comment]
@MatkovIvan Or you can use reflection too, there is an example in the updates branch to enable updates only in the project build flavour.

**Comment by guperrot** [comment]
Please run `./gradlew build coverageReport` locally to debug why the build fails. When it succeeds, use the coverage report in the `sdk/build` local folder to debug what tests are missing.

File: sdk/mobile-center-push/build.gradle [hunk 1]
```diff
@@ -0,0 +1,10 @@
+//
+// :sdk:push
+//
+project.description = 'This package contains functionalities to recive push notifications for your application.'
+evaluationDependsOn(':sdk')
+
+dependencies {
+    compile project(':sdk:mobile-center')
+    compile 'com.google.android.gms:play-services-gcm:10.2.0'
```
Comment (guperrot) [inline sdk/mobile-center-push/build.gradle]: GCM is deprecated, we have to use Firebase SDK now.

File: sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/Push.java [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+     */
+    private static final String PREFERENCE_PREFIX = SERVICE_NAME + ".";
+
+    /**
+     * Preference key to store push token.
+     */
+    @VisibleForTesting
+    static final String PREFERENCE_KEY_PUSH_TOKEN = PREFERENCE_PREFIX + "push_token";
+
+
```
Comment (guperrot) [inline sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/Push.java]: Extra blank line that should be removed here.

File: sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/Push.java [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    /**
+     * Shared instance.
+     */
+    @SuppressLint("StaticFieldLeak")
+    private static Push sInstance = null;
```
Comment (guperrot) [inline sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/Push.java]: In our coding conventions we don't initiliaze like that when its null, 0 or false.

File: sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/Push.java [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+     * Enqueue a push installation log.
+     *
+     * @param pushToken the push token value
+     */
+    private void enqueuePushInstallationLog(@NonNull String pushToken) {
+        if (isInactive())
+            return;
+        if (mPushTokenSent)
+            return;
+
```
Comment (guperrot) [inline sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/Push.java]: Either add a comment before each block or remove blank line.

File: sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/Push.java [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+        if (mPushToken != null) {
+            enqueuePushInstallationLog(mPushToken);
+        } else if (mSenderId != null) {
+            final PushTokenTask pushTokenTask = new PushTokenTask();
+            try {
+                pushTokenTask.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
+            } catch (final RejectedExecutionException e) {
+                MobileCenterLog.error(LOG_TAG, "THREAD_POOL_EXECUTOR saturated", e);
```
Comment (guperrot) [inline sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/Push.java]: Please add a TODO comment here, so that we use ASyncTaskUtils when its gets merged in the future.

File: sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/Push.java [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+    private synchronized void setInstanceSenderId(String senderId) {
+        mSenderId = senderId;
+    }
+
+    @VisibleForTesting
+    class PushTokenTask extends AsyncTask<Void, Void, String> {
+
+        @Override
+        protected String doInBackground(Void[] params) {
+            InstanceID instanceID = InstanceID.getInstance(mContext);
```
Comment (guperrot) [inline sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/Push.java]: GCM being deprecated, this code has to use Firebase SDK now.

File: sdk/mobile-center-push/src/test/java/com/microsoft/azure/mobile/push/PushTest.java [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.powermock.api.mockito.PowerMockito.mockStatic;
+import static org.powermock.api.mockito.PowerMockito.verifyStatic;
+import static org.powermock.api.mockito.PowerMockito.when;
+import static org.powermock.api.mockito.PowerMockito.whenNew;
+
+@SuppressWarnings("unused")
+@RunWith(PowerMockRunner.class)
```
Comment (guperrot) [inline sdk/mobile-center-push/src/test/java/com/microsoft/azure/mobile/push/PushTest.java]: Ok that's the reason why code coverage fails to report percentage for tests originating from this class: if you use `@RunWith(PowerMockRunner.class)`, it will not cover classes that are in `@PrepareForTest`.

The solution for that is to remove `@RunWith(PowerMockRunner.class)` and instead use:

```
    @Rule
    public PowerMockRule mPowerMockRule = new PowerMockRule();
```

The rule is slower and requires the vm arg `-noverify` to test from studio so that's why we don't use it everywhere, but here you need the rule.

File: sdk/mobile-center-push/src/test/java/com/microsoft/azure/mobile/push/PushTest.java [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    public void handlePushToken() {
+        String testToken = "TEST";
+        Push push = Push.getInstance();
+        Channel channel = mock(Channel.class);
+        push.onChannelReady(mock(Context.class), channel);
+        push.handlePushToken(testToken);
+        verifyStatic(times(1));
+        StorageHelper.PreferencesStorage.putString(eq(PREFERENCE_KEY_PUSH_TOKEN), eq(testToken));
+
+        // For check enqueue only once
```
Comment (guperrot) [inline sdk/mobile-center-push/src/test/java/com/microsoft/azure/mobile/push/PushTest.java]: We use // only to disable code temporarily.
Real comments are /*

File: sdk/mobile-center-push/build.gradle [hunk 9]
```diff
@@ -0,0 +1,10 @@
+//
```
Comment (jaelim-ms) [inline sdk/mobile-center-push/build.gradle]: Remove the first 3 lines.

Comment (MatkovIvan) [inline sdk/mobile-center-push/build.gradle]: @jaelim-ms I used other modules as example. In analytics and crashes it is same.  So, why here must be different?

Comment (guperrot) [inline sdk/mobile-center-push/build.gradle]: We cleaned that in updates actually. Reason is that old Android Studio was showing build.gradle as editor title so it was confusing. Now it's showing the module name in editor tab title so we don't need this comment anymore.

File: apps/sasquatch/src/main/java/com/microsoft/azure/mobile/sasquatch/activities/SettingsActivity.java [hunk 10]
```diff
@@ -80,7 +80,7 @@ public boolean isEnabled() {
                 @Override
                 public void setEnabled(boolean enabled) {
                     Crashes.setEnabled(enabled);
-                    crashesEnabledPreference.setChecked(Analytics.isEnabled());
+                    crashesEnabledPreference.setChecked(Crashes.isEnabled());
```
Comment (guperrot) [inline apps/sasquatch/src/main/java/com/microsoft/azure/mobile/sasquatch/activities/SettingsActivity.java]: Good catch :D Will backport to develop ;)

Comment (jaelim-ms) [inline apps/sasquatch/src/main/java/com/microsoft/azure/mobile/sasquatch/activities/SettingsActivity.java]: @guperrot fixed this on `develop` branch. Revert this change, merging develop will take care of this change.

Comment (guperrot) [inline apps/sasquatch/src/main/java/com/microsoft/azure/mobile/sasquatch/activities/SettingsActivity.java]: Just merging develop is enough, no need to revert I think.

Comment (jaelim-ms) [inline apps/sasquatch/src/main/java/com/microsoft/azure/mobile/sasquatch/activities/SettingsActivity.java]: The same change will have different commit IDs. If it doesn't matter, I don't care. 😄

Comment (guperrot) [inline apps/sasquatch/src/main/java/com/microsoft/azure/mobile/sasquatch/activities/SettingsActivity.java]: I think it does not matter, and git annotate will likely show the one from develop after merge.

File: sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/Push.java [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+     * Handle push token update success.
+     *
+     * @param token the push token value
+     */
+    @VisibleForTesting
+    synchronized void onTokenRefresh(@NonNull String token) {
+        mPushToken = token;
+        MobileCenterLog.debug(LOG_TAG, "Push token: " + mPushToken);
+        PreferencesStorage.putString(PREFERENCE_KEY_PUSH_TOKEN, mPushToken);
+        enqueuePushInstallationLog(mPushToken);
```
Comment (guperrot) [inline sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/Push.java]: If the token is refreshed but we already sent one during process lifetime we will skip sending which is incorrect behavior. It's not unusual that Google can replace the token with a "canonical" version so we should be ready to accept a new token any time. I would replace the boolean check by a string comparison with previous token value.

File: sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/Push.java [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+        super.onChannelReady(context, channel);
+        updatePushToken();
+    }
+
+    @Override
+    public synchronized void setInstanceEnabled(boolean enabled) {
+        super.setInstanceEnabled(enabled);
+        if (enabled) {
+            updatePushToken();
+        }
```
Comment (guperrot) [inline sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/Push.java]: The rule for modules when disabled is to reset any state, delete persisted state also.
Expected behavior when we disable and re-enable is to get token and send token again.

File: sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/Push.java [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+     * @param enabled current state.
+     */
+    private synchronized void applyEnabledState(boolean enabled) {
+        if (enabled && mChannel != null) {
+            String token = FirebaseInstanceId.getInstance().getToken();
+            if (token != null) {
+                onTokenRefresh(token);
+            }
+        } else {
+            /* Reset module state if disabled */
```
Comment (guperrot) [inline sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/Push.java]: Missing blank line before comment

**Comment by guperrot** [comment]
Just waiting on missing code coverage now.
